### DISCUSSION TITLE
Visualize intermediate results of event iterations

### DISCUSSION THREAD

**Discussion by felix-marsollier2** [OpenModelica/OpenModelica#8481] [General]
### Description

The built-in "edge" function seems to not work properly. The output of components "Edge" and "FallingEdge" from the Modelica.Blocks.Logical package and the components "RisingEdge" and "FallingEdge" from from the Modelica.Blocks.MathBoolean package remain to its start value even if the input receive an edge.

It does seems to be a plotting effect that would hide the pic. I have tried to connect the output of a component using the edge function to an other component to ensure that the output is wrong at and after the event.

### Steps to Reproduce

The simple way to repreduce the error is to run the example "BooleanNetwork1" from the Modelica.Blocks.Examples package and monitor the result "y" from the "rising" and "falling" components.

### Expected Behavior

The output should be true when an edge is applied to the input 

### Version and OS
 - OpenModelica Version: OpenModelica v1.18.0 (64-bit)
 - OS: Windows 10, 64 bit
 - Versions of used Modelica libraries if applicable: 3.2.3

**Comment by casella** [comment]
@felix-marsollier2, `edge(b)` is defined as `b and not pre(b)`, and as such can only remain true for an infintesimally short amount of time during event iteration. Consider this example
```
model M
  Boolean b(start = false, fixed = true), e;
equation
  when time > 0.5 then 
    b = true;
  end when;
  e = edge(b);
end M;
```
The simulation starts, `b` and `pre(b)` are both false , so `e` is also false until the event is triggered at time = 0.5.

When the event is triggered, the when equation is activated and the whole system is solved, including both equations. This means `b = true` and `e = b and not pre(b) = true and not false = true`. But now, as explained in [Appendix B](https://specification.modelica.org/maint/3.5/modelica-dae-representation.html) of the Modelica Specification, event iteration starts, so that it is possible to process other events that may be triggered from this first one. For that purpose, all pre-values of discrete values are set to the current discrete variable values, and then the active equations are solved again - the iteration stops when `v = pre(v)` for all variables after the last iteration. That's the key point, as you'll see in a second.

So, now `pre(b) = b = true`, `pre(e) = e = true`. The first equation is no longer active because the event has already been processed, hence we only need to solve the second equation, giving `e = b and not pre(b) = true and not true = false`. At this point `b = pre(b), but e <> pre(e)`, so we set `pre(b) = b = true` and `pre(e) = e = false` and iterate again.

When we solve the only active equation again, `e = b and not pre(b) = true and not true = false`. Now `e = pre(e)` also, so event iteration is over and we can restart the continuous-time integration.

As you can see, because of the even iteration terminating condition, it is impossible that the output of an `edge()` operator is true at the end of the even iteration process. The only role `edge()` can play is to trigger other events, that may eventually have permanently visible effects when the continuous simulation restarts, but its output can only be true during event iteration.

Unfortunately, Modelica tools (including OpenModelica and Dymola, which I also checked) do not store the results during event iterations - they only store the results when the event is first triggered and when the event process is over. Hence, the pulse remains "hidden". I'm afraid we can't do much about it, unless we decide that we actually want to also (optionally?) store the results of each event iteration. 

@kabdelhak, @AnHeuermann, @Adrpo, @sjoelund, what do you think?

**Comment by sjoelund** [comment]
I thought we did save the event iterations in the result-file. It's just not possible to plot them. If we don't store all of them out should be possible to do so on a flag. We still wouldn't be able to plot the values though.

**Comment by casella** [comment]
Why not? We already plot the values immediately before the event is processed and at the end of it - both have the same time, and a vertical line ensues if they are different. If we have more than than two ordered values at the same time event, we can just plot them in the same way. 

This at least would make the glitch of `edge()` visible.

Of course that wouldn't help sorting out exactly what happens during event iterations, because all these vertical lines would be superimposed. Unless we figure out a way to represent super-dense time.

**Comment by felix-marsollier2** [comment]
Ok I see. My apologize because my RsFlipFlop, to check whether edge had
been true during the simulation, wasn't set properly.
Maybe a small explaination in the component information could help in
components using the edge function.

With your explainations how sample function does work to be visible in the
result file ? From what I understand the output is true only during the
event and remains false during the continuous-time integration

Thanks a lot for your lights and thanks for the OpenModelica sofware.

Le ven. 1 oct. 2021 à 23:57, Francesco Casella ***@***.***> a
écrit :

**Comment by casella** [comment]
> With your explainations how sample function does work to be visible in the result file ?

You would get three rows with the same time value in the first column and with three different values of the `edge()` output: one false, one true, and then another one false.

**Comment by AJamesPhillips** [comment]
@felix-marsollier2 are you happy with closing this now?  (I can't do it for you, but I'm asking incase this issue is now resolved for you).

**Comment by casella** [comment]
@AJamesPhillips we could still change the runtime so that the spike in `e` is visible, so I left this open
